Add ELB resource

sahil-lakhwani opened the pull request:
Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

### Description of your changes

AWS offers 3 different types of load balancers. 

This PR adds a one of them - [Classic Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/introduction.html)

Classic Load Balancer is commonly called as ELB hence the name of the custom resource is also `ELB`.
This PR also adds a resource `ELBAttachment`  which manages the attachment of EC2 instances to a Classic Load Balancer.

Example manifest of both the resources can be found under `examples/elasticloadbalancing/`
### Checklist
<!--
Please run through the below readiness checklist. The first two items are
relevant to every Crossplane pull request.
-->
I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the dependencies in [`app.yaml`] to include any new role permissions.

[documentation]: https://github.com/crossplane/crossplane/tree/master/docs
[examples]: https://github.com/crossplane/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplane/crossplane/tree/master/PendingReleaseNotes.md
[`app.yaml`]: https://github.com/crossplane/provider-aws/blob/master/config/package/manifests/app.yaml

arush-sal commented on pkg/clients/elasticloadbalancing/elb/elb.go:
```diff
@@ -0,0 +1,174 @@
+
+// IsELBNotFound returns true if the error is because the item doesn't exist.
+func IsELBNotFound(err error) bool {
+	if awsErr, ok := err.(awserr.Error); ok {
```
```suggestion
	if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == ELBNotFound {
```

arush-sal commented on pkg/clients/elasticloadbalancing/elb/elb.go:
```diff
@@ -0,0 +1,174 @@
+
+// IsELBNotFound returns true if the error is because the item doesn't exist.
+func IsELBNotFound(err error) bool {
+	if awsErr, ok := err.(awserr.Error); ok {
+		if awsErr.Code() == ELBNotFound {
```
```suggestion
	if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == ELBNotFound {
```

negz commented:
Hi @sahil-lakhwani, can you please comment here and mark the PR ready for review when someone from your team has reviewed it and you're ready for @hasheddan to review. Thanks!

sahil-lakhwani commented:
@arush-sal I have addressed your comment. Also I have pushed some more code around updating the ELB resource - the list fields under `ELBParameters` needed to be sorted before a comparison. 
You may want to take a look on the added code.

negz commented:
@sahil-lakhwani Thanks for marking this ready for review. @arush-sal I'm going to wait for you to approve this PR before we move forward with a review from the maintainers.

rahulchheda commented on apis/elasticloadbalancing/v1alpha1/elb_attachment_types.go:
```diff
@@ -0,0 +1,74 @@
+type ELBAttachmentParameters struct {
+	// Name of the Elastic Load Balancer to which the instances will attach.
+	// +optional
+	ELBName string `json:"elbName,omitempty"`
```
Can we add a reference for this resource?

sahil-lakhwani replied:
Thanks for the catch, added the reference.

rahulchheda commented on apis/elasticloadbalancing/v1alpha1/elb_attachment_types.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+// ELBAttachmentObservation keeps the state for the external resource
+type ELBAttachmentObservation struct {
```
I guess, we can some fields in this resource, and respective logic to create Observation

sahil-lakhwani replied:
I didn't find anything suitable field to add to this.

rahulchheda commented on apis/elasticloadbalancing/v1alpha1/elb_types.go:
```diff
@@ -0,0 +1,175 @@
+	// The protocol to use for routing traffic to instances: HTTP, HTTPS, TCP, or
+	// SSL.
+	// +optional
+	InstanceProtocol *string `json:"instanceProtocol,omitempty"`
```
As this is optional, ca we add a comment, what is the default Protocol chosen?

rahulchheda commented on apis/elasticloadbalancing/v1alpha1/elb_types.go:
```diff
@@ -0,0 +1,175 @@
+	// or SSL.
+	Protocol string `json:"protocol"`
+
+	// The Amazon Resource Name (ARN) of the server certificate.
```
Can we change this comment to something like: `Sets the certificate that terminates the specified listener's SSL connections. The specified certificate replaces any prior certificate that was used on the same load balancer and port.`

sahil-lakhwani replied:
I think this statement is more for defining an action rather an entity/field. I haven't changed the comment for now.

rahulchheda replied:
Fair. Got it.

rahulchheda commented on apis/elasticloadbalancing/v1alpha1/elb_types.go:
```diff
@@ -0,0 +1,175 @@
+
+	// The type of a load balancer. Valid only for load balancers in a VPC.
+	// +optional
+	// +immutable
```
I guess you will have to remove this comment

sahil-lakhwani replied:
AFAIK, a field can be both `optional` and `immutable` at the same time.

rahulchheda replied:
Yup, saw the use-case as well

rahulchheda commented on apis/elasticloadbalancing/v1alpha1/elb_types.go:
```diff
@@ -0,0 +1,175 @@
+// +kubebuilder:printcolumn:name="READY",type="string",JSONPath=".status.conditions[?(@.type=='Ready')].status"
+// +kubebuilder:printcolumn:name="SYNCED",type="string",JSONPath=".status.conditions[?(@.type=='Synced')].status"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:subresource:status
```
Suggestion: Should we add more fields such as VPCID, or DNSName, directly in printcolumn?

rahulchheda commented on apis/elasticloadbalancing/v1alpha1/register.go:
```diff
@@ -0,0 +1,62 @@
+*/
+// +kubebuilder:object:generate=true
+// +groupName=elasticloadbalancing.aws.crossplane.io
+// +versionName=v1alpha1
```
I think we can more this doc.go, then we don't need to add it here.

rahulchheda commented on pkg/clients/elasticloadbalancing/elb/elb.go:
```diff
@@ -0,0 +1,223 @@
+)
+
+// A Client handles CRUD operations for Elastic Load Balancing resources.
+type Client elasticloadbalancingiface.ClientAPI
```
Rather than this, can we create an `interface`, with all the functions used in it, and then initiate that interface type?
Generally, I saw this practice, any reason to choose otherwise?

sahil-lakhwani replied:
I noticed there was already an interface in the AWS SDK and found it a bit easy to use than to create an interface. 
Doing this does expose all of the methods in `elasticloadbalancingiface.ClientAPI` to ELB controller which I don't see as a problem.

rahulchheda replied:
Seems good.

rahulchheda commented on examples/elasticloadbalancing/elb.yaml:
```diff
@@ -0,0 +1,24 @@
+apiVersion: elasticloadbalancing.aws.crossplane.io/v1alpha1
+kind: ELB
+metadata:
+  name: "sample-elb"
```
nit: Can we remove the double inverted commas

negz commented:
This looks pretty close to done to me; most of my comments were fairly minor. Can you please resolve any conversations that you feel are address and also comment here to let me know when it's ready for another look? I'll be offline next week, so I'd like to get this merged by Friday if possible.

sahil-lakhwani commented on pkg/clients/elasticloadbalancing/elb/elb_test.go:
```diff
@@ -0,0 +1,354 @@
+	}
+}
+
+func TestCreatePatch(t *testing.T) {
```
I have added a test case. 
If I understand the second question correctly - the `CreatePatch` method is not modifying the passed parameters.

sahil-lakhwani commented on pkg/controller/elasticloadbalancing/elb/controller.go:
```diff
@@ -0,0 +1,394 @@
+		return managed.ExternalUpdate{}, errors.Wrap(resource.Ignore(elb.IsELBNotFound, err), errDescribeTags)
+	}
+
+	patch, err := elb.CreatePatch(observed, cr.Spec.ForProvider, tagsResponse.TagDescriptions[0].Tags)
```
So `CreatePatch` here is being used to figure out what fields of the ELB need an update. This is done because there's no single API from AWS SDK to modify an ELB. (Added a comment  in the code as well)

Let me know if I understood your comment and answered it correctly.

sahil-lakhwani commented:
> This looks pretty close to done to me; most of my comments were fairly minor. Can you please resolve any conversations that you feel are address and also comment here to let me know when it's ready for another look? I'll be offline next week, so I'd like to get this merged by Friday if possible.

Thanks for the review @negz . I have mad changes to the code as per the comments and replied to the ones regarding `CreatePatch`.
